be/src/formats/parquet/file_writer.cpp

```diff
@@ -177,136 +191,645 @@ void ParquetBuildHelper::build_compression_type(::parquet::WriterProperties::Bui
+                                                  const ::parquet::schema::NodePtr& node, const ColumnPtr& col) {
+    unsigned char* nulls = nullptr;
+    if (col->is_nullable()) {
+        const auto null_column = down_cast<NullableColumn*>(col.get())->null_column();
+        nulls = null_column->get_data().data();
+    }
+
+    const auto data_column = ColumnHelper::get_data_column(col.get());
+    auto raw_col = down_cast<const RunTimeColumnType<TYPE_DATETIME>*>(data_column)->get_data().data();
+
+    _generate_rg_writer();
+    auto col_writer = down_cast<::parquet::Int64Writer*>(_rg_writer->column(_col_idx));
+    DCHECK(col_writer != nullptr);
+
+    std::vector<int64_t> values;
+    values.reserve(ctx.size());
+    for (auto i = 0; i < ctx.size(); i++) {
+        auto [idx, def_level, rep_level] = ctx.get(i);
+        if (idx == Context::kNULL || (nulls != nullptr && nulls[idx])) {
+            values.push_back(-1);
+            continue;
+        }
+        int64_t milliseconds = raw_col[idx].to_unix_second() * 1000;
+        values.push_back(milliseconds);
+    }
+
+    if (node->is_required()) {
+        col_writer->WriteBatch(ctx.size(), ctx._def_levels.data(), ctx._rep_levels.data(), values.data());
+    } else {
+        std::for_each(ctx._def_levels.begin(), ctx._def_levels.end(), [](int x) { return x + 1; });
+        col_writer->WriteBatch(ctx.size(), ctx._def_levels.data(), ctx._rep_levels.data(), values.data());
+        std::for_each(ctx._def_levels.begin(), ctx._def_levels.end(), [](int x) { return x - 1; });
     }
 
+    _buffered_values_estimate[_col_idx] = col_writer->EstimatedBufferedValueBytes();
+    _col_idx++;
+    return Status::OK();
+}
+
+template <LogicalType lt, ::parquet::Type::type pt>
+Status FileWriterBase::_add_int_column_chunk(const Context& ctx, const TypeDescriptor& type_desc,
+                                             const ::parquet::schema::NodePtr& node, const ColumnPtr& col) {
+    unsigned char* nulls = nullptr;
+    if (col->is_nullable()) {
+        const auto null_column = down_cast<NullableColumn*>(col.get())->null_column();
+        nulls = null_column->get_data().data();
+    }
+
+    const auto data_column = ColumnHelper::get_data_column(col.get());
+    const auto raw_col = down_cast<RunTimeColumnType<lt>*>(data_column)->get_data().data();
+
+    _generate_rg_writer();
+    auto col_writer =
+            down_cast<::parquet::TypedColumnWriter<::parquet::PhysicalType<pt>>*>(_rg_writer->column(_col_idx));
+    DCHECK(col_writer != nullptr);
+
+    using physical_type = typename ::parquet::type_traits<pt>::value_type;
+    std::vector<physical_type> values;
+    values.reserve(ctx.size());
+    for (auto i = 0; i < ctx.size(); i++) {
+        auto [idx, def_level, rep_level] = ctx.get(i);
+        if (idx == Context::kNULL || (nulls != nullptr && nulls[idx])) {
+            values.push_back(-1);
+            continue;
+        }
+        values.push_back(raw_col[idx]);
```

Comment:
you will copy memory in here

Reply:
This is unavoidable since we cannot use `raw_col` directly.

Reply:
Fixed. No more value copying for identical types.